includes/services/ThemeManager.php

```diff
@@ -0,0 +1,143 @@
+            //         'type' => 'danger',
+            //         'message' => _t('THEME_MANAGER_THEME_FOLDER') .$this->theme. _t('THEME_MANAGER_NOT_FOUND'),
+            //     ]);
+            return false;
+        }
+
+        if (!((!$useFallbackTheme &&file_exists('custom/'.$filePath)) || file_exists($filePath))) {
+            $this->errorMessage = '<div class="alert alert-danger">'._t('THEME_MANAGER_SQUELETTE_FILE') .$this->squelette. _t('THEME_MANAGER_NOT_FOUND').'</div>';
+            // $this->errorMessage = $this->TemplateEngine->render('@templates\alert-message.twig', [
+            //         'type' => 'danger',
+            //         'message' => _t('THEME_MANAGER_SQUELETTE_FILE') .$this->squelette. _t('THEME_MANAGER_NOT_FOUND'),
+            //     ]);
+            return false;
+        }
+        $filePath = (!$useFallbackTheme && file_exists('custom/'.$filePath)) ? 'custom/'. $filePath : $filePath;
+
+        $fileContent = file_get_contents($filePath);
+        if ($fileContent === false) {
+            $this->errorMessage = '<div class="alert alert-danger">'._t('THEME_MANAGER_ERROR_GETTING_FILE') .$filePath.'</div>';
+            // $this->errorMessage = $this->TemplateEngine->render('@templates\alert-message.twig', [
+            //     'type' => 'danger',
+            //     'message' => _t('THEME_MANAGER_ERROR_GETTING_FILE') .$filePath,
+            // ]);
+            return false;
+        }
+        $this->fileContent = $fileContent;
+        $this->fileLoaded = true;
+
+        // On recupere la partie haut du template et on execute les actions wikini
+        $templateCut = explode("{WIKINI_PAGE}", $fileContent);
+        $this->templateHeader = $templateCut[0] ?? '';
+        $this->templateFooter = (count($templateCut) > 0) ? $templateCut[1] : '';
+
+        return true;
+    }
+
+    public function getErrorMessage():string
+    {
+        return $this->errorMessage;
+    }
+
+    public function renderHeader():string
+    {
+        if ($this->fileLoaded || $this->loadTheme()) {
+            return $this->renderActions($this->templateHeader);
+        } else {
+            return '';
+        }
+    }
+    
+    public function renderFooter():string
+    {
+        if ($this->fileLoaded || $this->loadTheme()) {
+            return $this->renderActions($this->templateFooter);
+        } else {
+            return '';
+        }
+    }
+
+    private function renderActions(string $text):?string
+    {
+        if ($act = preg_match_all("/".'(\\{\\{)'.'(.*?)'.'(\\}\\})'."/is", $text, $matches)) {
+            $i = 0;
+            $j = 0;
+            foreach ($matches as $valeur) {
+                foreach ($valeur as $val) {
+                    if (isset($matches[2][$j]) && $matches[2][$j]!='') {
+                        $action = $matches[2][$j];
+                        $text = str_replace('{{'.$action.'}}', $this->Performer->run('action', 'formatter', ['text'=>'{{'.$action.'}}']), $text);
```

Comment:
je n'ai pas suivi trop ce que faisait le fix mais vu que tu en profites pour faire du refacto, je me permets de te demander pourquoi tu passes par le formatter plutôt qu'un $wiki->Format tel que c'était dans header.php :
` $this->Format('{{'.$action.'}}', 'action')` ?
au début j'ai trouvé moins lisible mais ensuite je me suis dit que ça devait être parce qu'il y a un peu moins de traitement en passant directement par le formatter "action" plutôt que le "wakka" appelé par `$this->Format`. Est-ce pour cela ?

Reply:
Initialement les actions header et footer appellent ceci : https://github.com/YesWiki/yeswiki/blob/25a162c7996fd0fda77ed507e2fc7721c934ce50/actions/header.php#L31

toutefois, l'action Format est maintenant refactorisée cf: https://github.com/YesWiki/yeswiki/blob/25a162c7996fd0fda77ed507e2fc7721c934ce50/includes/YesWiki.php#L749

Je me suis donc permis d'appeler directement le Performer dans le même état qu'avant par le formatter 'action'.

Effectivement, on perd un peu en lisibilité, est-ce que je dois remettre la fonction Format alors ?